src/Juvix/LLVM/Codegen/Graph.hs

```diff
@@ -0,0 +1,348 @@
+-- | Operations necessary to update nodes
+module Juvix.LLVM.Codegen.Graph where
+
+import Juvix.LLVM.Codegen.Block as Block
+import Juvix.LLVM.Codegen.Types as Types
+import Juvix.Library hiding (Type, local)
+import qualified Juvix.Utility.HashMap as Map
+import qualified LLVM.AST as AST
+import qualified LLVM.AST.Constant as C
+import qualified LLVM.AST.Name as Name
+import qualified LLVM.AST.Operand as Operand
+import qualified LLVM.AST.Type as Type
+
+-- TODO ∷ abstract out getElementPointer and load into a single operation
+
+--------------------------------------------------------------------------------
+-- Main Functions
+--------------------------------------------------------------------------------
+
+-- take 4 Operand.Operand
+link ∷
+  ( HasThrow "err" Errors m,
+    HasState "blockCount" Int m,
+    HasState "blocks" (Map.HashMap Name.Name BlockState) m,
+    HasState "count" Word m,
+    HasState "currentBlock" Name.Name m,
+    HasState "moduleDefinitions" [AST.Definition] m,
+    HasState "names" Names m,
+    HasState "symtab" (Map.HashMap Symbol Operand.Operand) m
+  ) ⇒
+  m Operand.Operand
+link = body >>= define Type.void "link" args
+  where
+    args =
+      ( [ (nodeType, "node_1"),
+          (numPorts, "port_1"),
+          (nodeType, "node_2"),
+          (numPorts, "port_2")
+        ] ∷
+          [(Type.Type, Name.Name)]
+      )
+    -- TODO ∷ Abstract most of the logic in this function
+    body = do
+      makeFunction "link" args
+      setPort ("node_1", "port_1") ("node_2", "port_2")
+      setPort ("node_2", "port_2") ("node_1", "port_1")
+      _ ← retNull
+      createBlocks
+
+-- perform offsets
+
+isBothPrimary = body >>= define Type.i1 "is_both_primary" args
+  where
+    args = [(nodeType, "node")]
+    body = do
+      makeFunction "is_both_primary" args
+      mainPort ← allocaNumPorts False (Operand.ConstantOperand (C.Int 32 0))
+      -- TODO ∷ Maybe bad, we should have an environemnt of edges that I can just call
+      edge ← findEdge
+      port ← call portType edge (Block.emptyArgs [mainPort])
+      nodePtr ← getElementPtr $
+        Types.Minimal
+          { Types.type' = nodePointer,
+            Types.address' = port,
+            Types.indincies' = Block.constant32List [0, 1]
+          }
+      node ← load nodePointer nodePtr
+      -- run compare function
+      cmp ← undefined
```

Comment:
This should just depend on the ports, right?

Reply:
The function works on a node, the compare should run that the node given back is the same as the current node. I'm thinking I need a reference to the node, and not the node itself to do this function?

From there I can convert PtrToInt.

Reply:
Ah I see. Yes, I think you will need a reference.